File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/utils/DBUtils.java
package com.firefly.example.reactive.coffee.store.utils;

import com.firefly.annotation.Component;
import com.firefly.annotation.Inject;
import com.firefly.utils.io.Resource;
import com.firefly.utils.log.slf4j.ext.LazyLogger;

import javax.sql.DataSource;
import java.sql.Connection;
import java.util.List;
import java.util.stream.Collectors;

import static com.firefly.db.init.ScriptUtils.executeSqlScript;

/**
 * @author Pengtao Qiu
 */
@Component
public class DBUtils {

    private static LazyLogger logger = LazyLogger.create();

    @Inject
    private DataSource dataSource;

    @Inject
    private ResourceUtils resourceUtils;

    public void createTables() {
        executeScript(getSchemaScript());
    }

    public void initializeData() {
        executeScript(getInitDataScript());
    }

    public void executeScript(Resource resource) {
        try (Connection connection = dataSource.getConnection()) {
            executeSqlScript(connection, resource);
        } catch (Exception e) {
            logger.error(() -> "execute SQL exception", e);
        }
    }


    public Resource getSchemaScript() {
        return resourceUtils.resource("/dbScript/coffee_store_schema.sql");
    }

    public Resource getInitDataScript() {
        return resourceUtils.resource("/dbScript/coffee_store_init_data.sql");
    }

    public static String toWildcard(List<?> list) {
        return list.parallelStream().map(o -> "?").collect(Collectors.joining(","));
    }

}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/utils/DBUtils.kt
package com.firefly.example.kotlin.coffee.store.utils

import com.firefly.annotation.Component
import com.firefly.annotation.Inject
import com.firefly.db.init.ScriptUtils.executeSqlScript
import com.firefly.kotlin.ext.log.Log
import com.firefly.kotlin.ext.log.error
import com.firefly.utils.io.Resource
import java.util.stream.Collectors
import javax.sql.DataSource

/**
 * @author Pengtao Qiu
 */
private val logger = Log.getLogger { }

@Component
class DBUtils {

    @Inject
    lateinit var dataSource: DataSource

    @Inject
    lateinit var resourceUtils: ResourceUtils

    val schemaScript: Resource
        get() = resourceUtils.resource("/dbScript/coffee_store_schema.sql")

    val initDataScript: Resource
        get() = resourceUtils.resource("/dbScript/coffee_store_init_data.sql")

    fun createTables() {
        executeScript(schemaScript)
    }

    fun initializeData() {
        executeScript(initDataScript)
    }

    fun executeScript(resource: Resource) = try {
        dataSource.connection.use { executeSqlScript(it, resource) }
    } catch (e: Exception) {
        logger.error({ "execute SQL exception" }, e)
    }

    companion object {
        fun toWildcard(list: List<*>): String {
            return list.parallelStream().map { "?" }.collect(Collectors.joining(","))
        }
    }

}